### ISSUE TITLE
TypeScript 3 Project References feature support

### ISSUE THREAD

**Issue by Stas-tolpekin** [angular/angular#37276]
### Bug Report or Feature Request (mark with an `x`)
```
- [ ] bug report -> please search issues before submitting
- [x] feature request
```

### Command (mark with an `x`)
```
- [ ] new
- [x] build
- [x] serve
- [x] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Versions
Node: 10.13.0
NPM: 3.10.10
Angular CLI: 7.0.4
Windows 10

### Desired functionality
I would like to use TypeScript 3 Project References feature in Angular 7 project (created by Angular CLI).

### Mention any other details that might be useful
More about TypeScript 3 Project References: https://www.typescriptlang.org/docs/handbook/project-references.html

**Comment by stephanmullerNL** [comment]
Hey folks, any update on this yet? It would be very useful indeed for large monorepo setups

**Comment by GEP-JOHNGOUDA** [comment]
Hi Folks, It would be great to have this feature. I just found myself in a situation where common code cannot be ported as a library and need the references feature to share code between multiple projects. It's been almost 3 years since the issue reported.

Thank you.

**Comment by sonntag-philipp** [comment]
Are there any updates on this?

**Comment by boginw** [comment]
This would help us immensely. Is there any update on this?

**Comment by BlindDespair** [comment]
This would be super helpful to us as well. I think this feature is a must to properly support monorepos.

**Comment by kepek** [comment]
Bump

**Comment by mparpaillon** [comment]
Thought my build was stuck but it was simply taking 3+ min to build in watch mode. Unusable DX. The /shared folder in my monorepo is quite big and I'm sure project reference would improve this greatly.

**Comment by jjgriff93** [comment]
STILL waiting for this

**Comment by stephanmullerNL** [comment]
It's been so long I actually stopped caring about this. I'm now a farmer

**Comment by Xriuk** [comment]
Bump

**Comment by olivier-spinelli** [comment]
This would be a real FEATURE for us.
Is it really as easy as having Angular calling "tsc -b" instead of "tsc"? 
Can we, at least, have a basic explanation of why this could be difficult?

**Comment by JonathanCabezas** [comment]
Can't believe nobody from the Angular team has commented on this lack of feature for 6 years...
Is Angular dead ?

This looks like a **blocking feature**.

**Comment by shaman-apprentice** [comment]
> i'd be very keen to see this sort of feature... the scenario we're looking at is something like this;
>```/root
>    /client
>    /server
>    /shared
> ```
> we've been looking at running typescript on the front end using Angular, a backend using [Nest.js](https://nestjs.com/) and > sharing all the models, utils etc in a third typescript project that can be imported by both the client and the server.

That is exactly my use case and I was surprised to find out that neither Angular nor NestJS currently support TypeScript's project references.

However, an easy workaround is to use [a file as npm package](https://docs.npmjs.com/cli/v8/commands/npm-install) - basically just write your */shared* as an npm package and in */client* and */server* you can add `"@myscope/shared": "../shared",` to your package json and run `npm install`. This will create a symlink, which works fine.

So I agree, that it's surprising that such a feature is not integrated in Angular (nor NestJS), where both have a strong focus on TypeScript. But for my 5 cents, I don't think it is a blocking issue.

**Comment by masonmark** [comment]
This issue is years old, but in the intervening years, TypeScript Project References have become a standard, required feature for performant TypeScript builds and reasonable language server performance with large codebases.

This is [documented on the TypeScript performance wiki](https://github.com/microsoft/TypeScript/wiki/Performance#using-project-references).

So I think it is definitely a blocking issue for _some_ projects, particularly large ones.

In my own experience, it is at least an _annoying_ problem — in a large heterogenous monorepo where only some of the apps are Angular, we aren't able to use normal TypeScript project references, because some of our apps use Angular. This can cause significant performance problems in editors like VS Code (I think all editors, though), where features like auto-complete, error-checking, and jump-to-definition are very slow or don't work.

There are workarounds, like moving Angular apps out of the monorepo, or making Angular apps consume only separately-built libraries, but AFAIK all the workarounds are costly (in terms of setup/maintenance/disruption).

**Comment by danielsharvey** [comment]
@masonmark Agree. Interestingly, Nx is moving to using workspaces and TypeScript Project References. I've asked about Angular there (Angular is well supported by Nx generally) and they say they are working on it but no details as yet.

See https://github.com/nrwl/nx/discussions/29099

**Comment by edbzn** [comment]
Note that Esbuild itself isn't supporting it (https://github.com/evanw/esbuild/issues/1250) I doubt Angular can integrate it without proper Esbuild support.

**Comment by masonmark** [comment]
If I understand the `esbuild` position, it is that this should be solved at a higher level than `esbuild` — the layer that is invoking `esbuild` (possibly with the help of a custom plugin). And there are some examples of people doing that for their own projects in that thread.

If you accept that stance, it implies that this is more of an Angular feature, then — especially if Angular wants to continue to use tools like `esbuild` and `vite` under the hood, while explictly [rejecting](https://github.com/angular/angular-cli/issues/26859) exposing the underlying config for those tools so that they can be extended. (The rationale for which I do understand, but the flip side of that coin is that if you don't let anybody else do anything, then you have to do it all  yourself.)

I guess Nx is trying to work around that — they have a long history of patching the Angular CLI and mucking around in its internals, to some extent — so maybe they are planning to do it at the monorepo-manager level.

But, it's notable that Nx 20.3 just shipped, expanding their (still experimental) [TypeScript Project References support](https://nx.dev/blog/nx-update-20-3#typescript-project-references-and-workspaces-support) to React, Vue, and Node apps — but notably, not Angular.

It's not clear to me if the Angular team has taken any kind of position on TS project references, or if they think the solution is "write your own custom builder". Having had to debug custom builder issues during the webpack era, this wouldn't be a road I'd be inclined to go down (as opposed to the less-satisfying but much simpler option of kicking the Angular apps out of the monorepo, and making them consume libs via private NPM packages (or similar)). 

But it might give folks who do have a commercial incentive to manage that complexity (like the Nx team) a way to do it. (?)

**Comment by HaasStefan** [comment]
Copying this from my duplicate issue:

I took a look into the compiler code and it seems like the project references are already correctly parsed in `readCommandLineAndConfiguration` in `compiler-cli/src/main.ts`, yet it is not returned by `readCommandLineAndConfiguration` and it is not passed into `NgTsc`. It just seems, as it would need to be drilled down to `ts.createProgram`. I am not sure, but as far as I can see, it should not require changes to `NgCompilerHost.wrap`

![Image](https://github.com/user-attachments/assets/70c1be02-b70d-43a5-9318-c031d52f1ccc)

@alxhub , I created a [Draft PR](https://github.com/angular/angular/pull/59897) and would be open to continue on this effort to put out a conforming PR if you agree with the direction of the changes to support project references.

**Comment by patriot-pledge** [comment]
It is absolutely APPALLING that six years later:

1. This has no comments from the Angular team
2. This is blocking massive projects from utilizing project references
3. THE PROPOSED SOLUTION IS 21 CHANGED LOC WHICH A NEW CONTRIBUTOR WAS ABLE TO FIGURE OUT?!

I literally might have to stop using angular CLI because it cannot handle project references. This issue has caused me dozens of hours of lost productivity, digging through stack overflow posts and github issues, running ng builds with unhelpful error messages.

### WHY DOES NG NOT SPIT OUT AN ERROR SAYING `composite: true` IS NOT A SUPPORTED OPTION?

**_I AM FURIOUS._**

**Comment by HaasStefan** [comment]
This response might be interesting to most of us as to why this is complicated to support in Angular https://github.com/angular/angular/pull/59897#issuecomment-2653642300

**Comment by zargham-leanix** [comment]
would be great to have an update on this topic from Angular and nx teams. We have a large monorepo and we are looking forward to enable npm workspaces + ts project references as outlined here: https://github.com/nrwl/nx/discussions/29099

**Comment by timtos** [comment]
Any updates on this? For large projects this is annoying. Thank you anyway for your good work in general!

**Comment by SergiiMost** [comment]
It's been almost 8 years—maybe it's time to finally fix this? :)

**Comment by lppedd** [comment]
In our Nx workspace, we'd like to migrate away from TS path aliases, to ease our lives when it will be the day we switch to TS 7, and to avoid TS transformers for aliases.

This is blocking us from performing this migration as using both references + aliases isn't a supported scenario for Nx, as far as I understand. But happy to be proven wrong.
